d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -797,33 +797,39 @@ void addWatcher(WildcardResourceWatcher watcher)
       }
     }
 
-    private void onData(String resourceName, ResourceUpdate data)
+    private void onData(String resourceName, ResourceUpdate update)
     {
-      if (Objects.equals(_data.get(resourceName), data))
+      if (Objects.equals(_data.get(resourceName), update))
       {
         _log.debug("Received resource update data equal to the current data. Will not perform the update.");
         return;
       }
       // null value guard to avoid overwriting the property with null
-      if (data != null && data.isValid())
+      if (update != null && update.isValid())
       {
-        _data.put(resourceName, data);
-        for (WildcardResourceWatcher watcher : _watchers)
-        {
-          watcher.onChanged(resourceName, data);
-        }
+        _data.put(resourceName, update);
       }
       else
       {
         if (_type == ResourceType.D2_URI_MAP || _type == ResourceType.D2_URI)
         {
-          RATE_LIMITED_LOGGER.warn("Received invalid data for {} {}, data: {}", _type, resourceName, data);
+          RATE_LIMITED_LOGGER.warn("Received invalid data for {} {}, data: {}", _type, resourceName, update);
         }
         else
         {
-          _log.warn("Received invalid data for {} {}, data: {}", _type, resourceName, data);
+          _log.warn("Received invalid data for {} {}, data: {}", _type, resourceName, update);
         }
       }
+
+      if (_data.get(resourceName) == null)
+      {
+        _log.info("Initializing {} {} to empty data.", _type, resourceName);
+        _data.put(resourceName, _type.emptyData());
+      }
```

Comment:
Why are you initializing it at all? Shouldn't it just be ignored if it's `null`, instead of notifying the watcher?

Reply:
This actually plays into this comment: https://github.com/linkedin/rest.li/pull/1038#discussion_r1919413458
If we have a new API for the wildcard subscriber then we don't have to do this. I don't think it really makes sense to add this in the map, otherwise new subscribers, which call `addWatcher`, will be notified of a resource that doesn't exist:
```java
    void addWatcher(WildcardResourceWatcher watcher)
    {
      _watchers.add(watcher);
      for (Map.Entry<String, ResourceUpdate> entry : _data.entrySet())
      {
        watcher.onChanged(entry.getKey(), entry.getValue());
        _log.debug("Notifying watcher of current data for resource {} of type {}: {}",
            entry.getKey(), _type, entry.getValue());
      }
    }
```

Reply:
"new subscribers, which call addWatcher, will be notified of a resource that doesn't exist" ---- this is exactly what this is supposed to do, to fail the request fast when the resource has bad data responded rather than keeps them waiting.

Reply:
And even having a new wildcard subscriber API which says "done with all updates" won't help, since the problematic resource has invalid data, if we don't put something in the map, even bus won't be updated and traffic request will hang until timing out. 
Image subscribing to d2 services with only wildcard, a call to d2://FooService won't get anything in the load balancer state if FooService has bad data returned by observer. It's not happening today because FooService is subscribed via individual subscriber, which add the empty holder when received bad data. A second call to d2://FooService will also fail immediately since event bus has a entry for it already (empty holder though).